File: modules/system/system.go
package system

import (
	"fmt"
	"os"
	"runtime"
	"strconv"
	"strings"
	"time"
)

type System struct{}

func (System) GetData(cfg map[string]any) []string {
	var out []string

	disksRaw, _ := cfg["disks"].([]any)
	for _, d := range disksRaw {
		dm, ok := d.(map[string]any)
		if !ok {
			continue
		}
		path, _ := dm["path"].(string)
		label, _ := dm["label"].(string)
		if path == "" {
			path = "/"
		}
		if label == "" {
			label = "Disk"
		}
		if line, err := diskUsage(path, label); err == nil {
			out = append(out, line)
		}
	}
	if len(disksRaw) == 0 {
		if line, err := diskUsage("/", "Root"); err == nil {
			out = append(out, line)
		}
	}

	out = append(out, ramUsage())

	cpuLine, tempLine := cpuStats()
	out = append(out, cpuLine)
	if tempLine != "" {
		out = append(out, tempLine)
	}

	out = append(out, uptimeStr())

	return out
}

func diskUsage(path, label string) (string, error) {
	total, used, err := statfs(path)
	if err != nil {
		return "", err
	}
	pct := float64(used) / float64(total) * 100
	return fmt.Sprintf("Disk %-6s : %.1f GB / %.1f GB  (%.1f%% used)",
		label,
		float64(used)/(1<<30),
		float64(total)/(1<<30),
		pct,
	), nil
}

func ramUsage() string {
	data, err := os.ReadFile("/proc/meminfo")
	if err != nil {
		return "RAM: Unavailable"
	}
	var total, avail int64
	for _, line := range strings.Split(string(data), "\n") {
		fields := strings.Fields(line)
		if len(fields) < 2 {
			continue
		}
		val, _ := strconv.ParseInt(fields[1], 10, 64)
		switch fields[0] {
		case "MemTotal:":
			total = val
		case "MemAvailable:":
			avail = val
		}
	}
	if total == 0 {
		return "RAM: Unavailable"
	}
	used := total - avail
	return fmt.Sprintf("RAM        : %.1f GB / %.1f GB  (%.1f%% used)",
		float64(used)/1048576, float64(total)/1048576, float64(used)/float64(total)*100)
}

func cpuStats() (cpuLine, tempLine string) {
	usage, err := cpuUsage()
	if err != nil {
		cpuLine = "CPU        : Unavailable"
	} else {
		cpuLine = fmt.Sprintf("CPU        : %.1f%%  (%d cores)", usage, runtime.NumCPU())
	}

	for _, zone := range []string{"thermal_zone0", "thermal_zone1", "thermal_zone2"} {
		path := "/sys/class/thermal/" + zone + "/temp"
		data, err := os.ReadFile(path)
		if err != nil {
			continue
		}
		val, err := strconv.ParseFloat(strings.TrimSpace(string(data)), 64)
		if err != nil {
			continue
		}
		tempLine = fmt.Sprintf("CPU Temp   : %.1f°C", val/1000)
		break
	}
	return
}

func cpuUsage() (float64, error) {
	read := func() (total, idle float64, err error) {
		data, err := os.ReadFile("/proc/stat")
		if err != nil {
			return
		}
		for _, line := range strings.Split(string(data), "\n") {
			if !strings.HasPrefix(line, "cpu ") {
				continue
			}
			fields := strings.Fields(line)[1:]
			var vals []float64
			for _, f := range fields {
				v, _ := strconv.ParseFloat(f, 64)
				vals = append(vals, v)
			}
			for _, v := range vals {
				total += v
			}
			if len(vals) > 3 {
				idle = vals[3]
			}
			return
		}
		err = fmt.Errorf("cpu line not found")
		return
	}

	t1, i1, err := read()
	if err != nil {
		return 0, err
	}
	time.Sleep(150 * time.Millisecond)
	t2, i2, err := read()
	if err != nil {
		return 0, err
	}

	dt := t2 - t1
	di := i2 - i1
	if dt == 0 {
		return 0, nil
	}
	return 100 * (1 - di/dt), nil
}

func uptimeStr() string {
	data, err := os.ReadFile("/proc/uptime")
	if err != nil {
		return ""
	}
	fields := strings.Fields(string(data))
	if len(fields) == 0 {
		return ""
	}
	secs, _ := strconv.ParseFloat(fields[0], 64)
	d := time.Duration(secs) * time.Second
	h := int(d.Hours())
	m := int(d.Minutes()) % 60
	return fmt.Sprintf("Uptime     : %dh %dm", h, m)
}

type Essentials struct{}

func (Essentials) GetData(_ map[string]any) []string {
	return []string{"⚡ Report generated. Have a great day!"}
}

File: modules/system/system.py
import os
import shutil
import time

def get_data(config):
    disks_input = config.get('disks', [("/", "OS")])
    results = []
    
    disks = []
    for item in disks_input:
        if isinstance(item, dict):
            disks.append((item.get('path', '/'), item.get('label', 'Disk')))
        elif isinstance(item, (list, tuple)) and len(item) >= 2:
            disks.append((item[0], item[1]))

    for path, label in disks:
        try:
            total, used, free = shutil.disk_usage(path)
            total_gb = total / (1024**3)
            used_gb = used / (1024**3)
            percent = (used / total) * 100
            results.append(f"Disk {label}: {used_gb:.1f} GB / {total_gb:.1f} GB ({percent:.1f}% used)")
        except Exception:
            results.append(f"Disk {path}: Not found")
            
    try:
        with open("/proc/meminfo", "r") as f:
            memlines = f.readlines()
            total_kb = int(memlines[0].split()[1])
            avail_kb = 0
            for line in memlines:
                if "MemAvailable" in line:
                    avail_kb = int(line.split()[1])
                    break
            if avail_kb == 0:
                avail_kb = int(memlines[1].split()[1])
                
            used_gb = (total_kb - avail_kb) / (1024 * 1024)
            total_gb = total_kb / (1024 * 1024)
            results.append(f"Ram usage: {used_gb:.1f} GB / {total_gb:.1f} GB")
    except Exception:
        results.append("Ram usage: Unavailable")
        
    try:
        def get_cpu_times():
            with open("/proc/stat", "r") as f:
                line = f.readline()
                fields = [float(column) for column in line.strip().split()[1:]]
                return sum(fields), fields[3]
        
        t1, i1 = get_cpu_times()
        time.sleep(0.1)
        t2, i2 = get_cpu_times()
        cpu_usage = 100 * (1 - (i2 - i1) / (t2 - t1))
        results.append(f"CPU Usage: {cpu_usage:.1f}%")
        
        temp = "N/A"
        for tzone in ['thermal_zone0', 'thermal_zone1']:
            tpath = f"/sys/class/thermal/{tzone}/temp"
            if os.path.exists(tpath):
                with open(tpath, "r") as f:
                    temp = f"{float(f.read()) / 1000.0:.1f}"
                    break
        results.append(f"CPU Temp: {temp}C")
    except Exception:
        results.append("CPU Stats: Unavailable")
        
    return results
